### ISSUE TITLE
Assert fail in regcomp.c with no other symptoms: perl -e '/(?<=/'

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#15332]
Migrated from [rt.perl.org#128170](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=128170) (status was 'resolved')

Searchable as RT128170$

**Comment by p5pRT** [comment]
### From @dcollinsn

Greetings Porters\,

I have compiled bleadperl with the afl\-gcc compiler using&#8203;:

\./Configure \-Dusedevel \-Dprefix='/usr/local/perl\-afl' \-Dcc='ccache afl\-gcc' \-Uuselongdouble \-Duse64bitall \-Doptimize=\-g \-Uversiononly \-Uman1dir \-Uman3dir \-Dusequadmath \-des
AFL\_HARDEN=1 make && make test

And then fuzzed the resulting binary using&#8203;:

AFL\_NO\_VAR\_CHECK=1 afl\-fuzz \-i in \-o out bin/perl @&#8203;@&#8203;

After reducing testcases using \`afl\-tmin\` and performing additional minimization by hand\, I have located the following testcase that triggers an assert fail in debugging builds of the perl interpreter\. The testcase is the 6\-character file&#8203;:

/\(?\<=/

On normal builds\, this exits with the expected error\. On debug builds\, this returns an assert fail\.

dcollins@&#8203;nightshade64&#8203;:\~$ perl/miniperl \-e '/\(?\<=/'
Unmatched \( in regex; marked by \<\-\- HERE in m/\( \<\-\- HERE ?\<=/ at \-e line 1\.
dcollins@&#8203;nightshade64&#8203;:\~$ perldebug/miniperl \-e '/\(?\<=/'
miniperl&#8203;: regcomp\.c&#8203;:10621&#8203;: S\_reg&#8203;: Assertion \`\(pRExC\_state\->parse\) \< \(pRExC\_state\->end\)' failed\.
Aborted

Debugging tool output is below\. A bisect was performed and points to the following diff&#8203;:

cfbef7dc3bfb89e4ed2c00ea9c9e3bcfd0b170fd is the first bad commit
commit cfbef7dc3bfb89e4ed2c00ea9c9e3bcfd0b170fd
Author&#8203;: Karl Williamson \<khw@&#8203;cpan\.org>
Date&#8203;:   Wed Feb 10 16&#8203;:27&#8203;:13 2016 \-0700

&nbsp;    regcomp\.c&#8203;: Fix some parsing glitches

&nbsp;    I undertook a code review of how regcomp\.c parses things in light of the
&nbsp;    tickets found by the fuzzer\,
&nbsp;    https://rt-archive.perl.org/perl5/Ticket/Display.html?id=126546.&nbsp;  This commit is the
&nbsp;    result of my efforts so far\.  I was not planning to push it now\, but the
&nbsp;    work found a couple of new error messages that should be raised\, and
&nbsp;    this has to be done before the visible changes code freeze coming up all
&nbsp;    too soon\.  I will add test cases after that freeze\, including if to see
&nbsp;    that these changes fix all the observed issues\.

&nbsp;    The audit was tedious\, and may have missed some things\.  Several issues
&nbsp;    occurred in multiple places\.  One is to not advance the parse by
&nbsp;    UTF8SKIP appropriately; another is to subtract one byte from the parse
&nbsp;    and assume that that is pointing to the beginning of the previous
&nbsp;    character \(which under UTF\-8 it may not\)\.  Another is to assume that
&nbsp;    that the pattern is a C string\, that there are no interior NULs in it\.
&nbsp;    I also found unnecessary tests\, given that an SV always has a
&nbsp;    terminating NUL\.

:040000 040000 a58a2d3154c5e346f9dfab14f7ca9897cbe06cfb 41b83ee5e555d3ddc87935c3a3b30db821741214 M      pod
:100644 100644 c00fdffd93dd9de69326e9b567d912dd9a1bee3f aa06bc6e8da12440f5d5b649cd5faa24a8853cc7 M      regcomp\.c
bisect run success



\*\*GDB\*\*

dcollins@&#8203;nightshade64&#8203;:\~$ gdb \-\-args perldebug/miniperl \-e '/\(?\<=/'
GNU gdb \(GDB\) 7\.10
Copyright \(C\) 2015 Free Software Foundation\, Inc\.
License GPLv3\+&#8203;: GNU GPL version 3 or later \<http&#8203;://gnu\.org/licenses/gpl\.html>
This is free software&#8203;: you are free to change and redistribute it\.
There is NO WARRANTY\, to the extent permitted by law\.  Type "show copying"
and "show warranty" for details\.
This GDB was configured as "x86\_64\-unknown\-linux\-gnu"\.
Type "show configuration" for configuration details\.
For bug reporting instructions\, please see&#8203;:
\<http&#8203;://www\.gnu\.org/software/gdb/bugs/>\.
Find the GDB manual and other documentation resources online at&#8203;:
\<http&#8203;://www\.gnu\.org/software/gdb/documentation/>\.
For help\, type "help"\.
Type "apropos word" to search for commands related to "word"\.\.\.
Reading symbols from perldebug/miniperl\.\.\.done\.
\(gdb\) run
Starting program&#8203;: /home/dcollins/perldebug/miniperl \-e /\\\(\\?\\\<=/
\[Thread debugging using libthread\_db enabled\]
Using host libthread\_db library "/lib/x86\_64\-linux\-gnu/libthread\_db\.so\.1"\.
miniperl&#8203;: regcomp\.c&#8203;:10621&#8203;: S\_reg&#8203;: Assertion \`\(pRExC\_state\->parse\) \< \(pRExC\_state\->end\)' failed\.

Program received signal SIGABRT\, Aborted\.
0x00007ffff6cf9478 in raise \(\) from /lib/x86\_64\-linux\-gnu/libc\.so\.6
\(gdb\) bt
\#0  0x00007ffff6cf9478 in raise \(\) from /lib/x86\_64\-linux\-gnu/libc\.so\.6
\#1  0x00007ffff6cfa8fa in abort \(\) from /lib/x86\_64\-linux\-gnu/libc\.so\.6
\#2  0x00007ffff6cf23a7 in ?? \(\) from /lib/x86\_64\-linux\-gnu/libc\.so\.6
\#3  0x00007ffff6cf2452 in \_\_assert\_fail \(\) from /lib/x86\_64\-linux\-gnu/libc\.so\.6
\#4  0x00000000006efe09 in S\_reg \(pRExC\_state=0x7fffffffdcf0\, paren=60\, flagp=0x7fffffffd8bc\, depth=5\) at regcomp\.c&#8203;:10621
\#5  0x00000000007186ed in S\_regatom \(pRExC\_state=pRExC\_state@&#8203;entry=0x7fffffffdcf0\, flagp=flagp@&#8203;entry=0x7fffffffda34\, depth=depth@&#8203;entry=4\) at regcomp\.c&#8203;:12354
\#6  0x0000000000725d0d in S\_regpiece \(depth=\<optimized out>\, flagp=\<synthetic pointer>\, pRExC\_state=0x7fffffffdcf0\) at regcomp\.c&#8203;:11421
\#7  S\_regbranch \(pRExC\_state=pRExC\_state@&#8203;entry=0x7fffffffdcf0\, flagp=flagp@&#8203;entry=0x7fffffffdacc\, first=first@&#8203;entry=1\, depth=depth@&#8203;entry=2\) at regcomp\.c&#8203;:11346
\#8  0x0000000000745097 in S\_reg \(pRExC\_state=pRExC\_state@&#8203;entry=0x7fffffffdcf0\, flagp=flagp@&#8203;entry=0x7fffffffdbbc\, depth=1\, paren=0\) at regcomp\.c&#8203;:11084
\#9  0x000000000077b231 in Perl\_re\_op\_compile \(patternp=\<optimized out>\, pat\_count=\<optimized out>\, expr=\<optimized out>\, eng=0xf33160 \<PL\_core\_reg\_engine>\, old\_re=0x0\,
&nbsp;    is\_bare\_re=\<optimized out>\, orig\_rx\_flags=0\, pm\_flags=0\) at regcomp\.c&#8203;:7027
\#10 0x00000000004c4033 in Perl\_pmruntime \(o=0x1202718\, expr=0x12026d8\, repl=0x0\, isreg=isreg@&#8203;entry=true\, floor=\<optimized out>\) at op\.c&#8203;:5632
\#11 0x000000000064b5cd in Perl\_yyparse \(gramtype=gramtype@&#8203;entry=258\) at perly\.y&#8203;:1028
\#12 0x0000000000512111 in S\_parse\_body \(xsinit=0xec1b00 \<xs\_init>\, env=0x0\) at perl\.c&#8203;:2365
\#13 perl\_parse \(my\_perl=\<optimized out>\, xsinit=xsinit@&#8203;entry=0xec1b00 \<xs\_init>\, argc=\<optimized out>\, argv=\<optimized out>\, env=env@&#8203;entry=0x0\) at perl\.c&#8203;:1681
\#14 0x000000000040f838 in main \(argc=\<optimized out>\, argv=\<optimized out>\, env=\<optimized out>\) at miniperlmain\.c&#8203;:120
\(gdb\) f 4
\#4  0x00000000006efe09 in S\_reg \(pRExC\_state=0x7fffffffdcf0\, paren=60\, flagp=0x7fffffffd8bc\, depth=5\) at regcomp\.c&#8203;:10621
10621                   assert\(RExC\_parse \< RExC\_end\);
\(gdb\) info locals
is\_logical = false
seqstart = \<optimized out>
endptr = 0x0
has\_intervening\_patws = false
ret = 0x0
br = \<optimized out>
lastbr = \<optimized out>
ender = 0x0
parno = 0
flags = 0
oregflags = 0
have\_branch = false
is\_open = false
freeze\_paren = 0
after\_freeze = 0
num = \<optimized out>
parse\_start = 0x1202861 "?\<="
oregcomp\_parse = 0x1202861 "?\<="
re\_debug\_flags = 0
\(gdb\) l
10616                       goto capturing\_parens;
10617                   \}
10618                   RExC\_seen |= REG\_LOOKBEHIND\_SEEN;
10619                   RExC\_in\_lookbehind\+\+;
10620                   RExC\_parse\+\+;
10621                   assert\(RExC\_parse \< RExC\_end\);
10622                   /\* FALLTHROUGH \*/
10623               case '='&#8203;:           /\* \(?=\.\.\.\) \*/
10624                   RExC\_seen\_zerolen\+\+;
10625                   break;
\(gdb\)


\*\*PERL \-V\*\*

dcollins@&#8203;nightshade64&#8203;:\~/perl$ \./perl \-Ilib \-V
Summary of my perl5 \(revision 5 version 25 subversion 1\) configuration&#8203;:
&nbsp;  Commit id&#8203;: 8255316ac82c0fbd7ae49f2f9a217ce063d7bbfb
&nbsp;  Platform&#8203;:
&nbsp;    osname=linux\, osvers=4\.5\.0\-2\-amd64\, archname=x86\_64\-linux\-ld
&nbsp;    uname='linux nightshade64 4\.5\.0\-2\-amd64 \#1 smp debian 4\.5\.3\-2 \(2016\-05\-08\) x86\_64 gnulinux '
&nbsp;    config\_args='\-Dusedevel \-Dprefix=/usr/local/perl\-afl \-Dcc=ccache afl\-gcc \-Duselongdouble \-Duse64bitall \-Doptimize=\-g \-Uversiononly \-Uman1dir \-Uman3dir \-des'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    useithreads=undef\, usemultiplicity=undef
&nbsp;    use64bitint=define\, use64bitall=define\, uselongdouble=define
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='ccache afl\-gcc'\, ccflags ='\-fwrapv \-fno\-strict\-aliasing \-pipe \-fstack\-protector\-strong \-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
&nbsp;    optimize='\-g'\,
&nbsp;    cppflags='\-fwrapv \-fno\-strict\-aliasing \-pipe \-fstack\-protector\-strong \-I/usr/local/include'
&nbsp;    ccversion=''\, gccversion='6\.1\.0'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=8\, ptrsize=8\, doublesize=8\, byteorder=12345678\, doublekind=3
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16\, longdblkind=3
&nbsp;    ivtype='long'\, ivsize=8\, nvtype='long double'\, nvsize=16\, Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=16\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='ccache afl\-gcc'\, ldflags =' \-fstack\-protector\-strong \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /usr/local/lib/gcc/x86\_64\-pc\-linux\-gnu/6\.1\.0/include\-fixed /usr/include/x86\_64\-linux\-gnu /usr/lib /lib/x86\_64\-linux\-gnu /lib/\.\./lib /usr/lib/x86\_64\-linux\-gnu /usr/lib/\.\./lib /lib
&nbsp;    libs=\-lpthread \-lnsl \-ldl \-lm \-lcrypt \-lutil \-lc
&nbsp;    perllibs=\-lpthread \-lnsl \-ldl \-lm \-lcrypt \-lutil \-lc
&nbsp;    libc=libc\-2\.22\.so\, so=so\, useshrplib=false\, libperl=libperl\.a
&nbsp;    gnulibc\_version='2\.22'
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-Wl\,\-E'
&nbsp;    cccdlflags='\-fPIC'\, lddlflags='\-shared \-g \-L/usr/local/lib \-fstack\-protector\-strong'


Characteristics of this binary \(from libperl\)&#8203;:
&nbsp;  Compile\-time options&#8203;: HAS\_TIMES PERLIO\_LAYERS PERL\_COPY\_ON\_WRITE
&nbsp;                        PERL\_DONT\_CREATE\_GVSV
&nbsp;                        PERL\_HASH\_FUNC\_ONE\_AT\_A\_TIME\_HARD PERL\_MALLOC\_WRAP
&nbsp;                        PERL\_OP\_PARENT PERL\_PRESERVE\_IVUV PERL\_USE\_DEVEL
&nbsp;                        USE\_64\_BIT\_ALL USE\_64\_BIT\_INT USE\_LARGE\_FILES
&nbsp;                        USE\_LOCALE USE\_LOCALE\_COLLATE USE\_LOCALE\_CTYPE
&nbsp;                        USE\_LOCALE\_NUMERIC USE\_LOCALE\_TIME USE\_LONG\_DOUBLE
&nbsp;                        USE\_PERLIO USE\_PERL\_ATOF
&nbsp;  Built under linux
&nbsp;  Compiled at May 16 2016 07&#8203;:45&#8203;:27
&nbsp;  @&#8203;INC&#8203;:
&nbsp;    lib
&nbsp;    /usr/local/perl\-afl/lib/site\_perl/5\.25\.1/x86\_64\-linux\-ld
&nbsp;    /usr/local/perl\-afl/lib/site\_perl/5\.25\.1
&nbsp;    /usr/local/perl\-afl/lib/5\.25\.1/x86\_64\-linux\-ld
&nbsp;    /usr/local/perl\-afl/lib/5\.25\.1
&nbsp;    /usr/local/perl\-afl/lib/site\_perl/5\.24\.0
&nbsp;    /usr/local/perl\-afl/lib/site\_perl
&nbsp;    \.

**Comment by p5pRT** [comment]
### From @demerphq

On 17 May 2016 3&#8203;:04 p\.m\.\, "Dan Collins" \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
>
> \# New Ticket Created by  Dan Collins
> \# Please include the string&#8203;:  \[perl \#128170\]
> \# in the subject line of all future correspondence about this issue\.
> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=128170 >
>
>
> Greetings Porters\,
>
> I have compiled bleadperl with the afl\-gcc compiler using&#8203;:
>
> \./Configure \-Dusedevel \-Dprefix='/usr/local/perl\-afl' \-Dcc='ccache
afl\-gcc' \-Uuselongdouble \-Duse64bitall \-Doptimize=\-g \-Uversiononly
\-Uman1dir \-Uman3dir \-Dusequadmath \-des
> AFL\_HARDEN=1 make && make test
>
> And then fuzzed the resulting binary using&#8203;:
>
> AFL\_NO\_VAR\_CHECK=1 afl\-fuzz \-i in \-o out bin/perl @&#8203;@&#8203;
>
> After reducing testcases using \`afl\-tmin\` and performing additional
minimization by hand\, I have located the following testcase that triggers
an assert fail in debugging builds of the perl interpreter\. The testcase is
the 6\-character file&#8203;:
>
> /\(?\<=/
>
> On normal builds\, this exits with the expected error\. On debug builds\,
this returns an assert fail\.
>
> dcollins@&#8203;nightshade64&#8203;:\~$ perl/miniperl \-e '/\(?\<=/'
> Unmatched \( in regex; marked by \<\-\- HERE in m/\( \<\-\- HERE ?\<=/ at \-e line
1\.
> dcollins@&#8203;nightshade64&#8203;:\~$ perldebug/miniperl \-e '/\(?\<=/'
> miniperl&#8203;: regcomp\.c&#8203;:10621&#8203;: S\_reg&#8203;: Assertion \`\(pRExC\_state\->parse\) \<
\(pRExC\_state\->end\)' failed\.
> Aborted
>
> Debugging tool output is below\. A bisect was performed and points to the
following diff&#8203;:
>
> cfbef7dc3bfb89e4ed2c00ea9c9e3bcfd0b170fd is the first bad commit
> commit cfbef7dc3bfb89e4ed2c00ea9c9e3bcfd0b170fd
> Author&#8203;: Karl Williamson \<khw@&#8203;cpan\.org>
> Date&#8203;:   Wed Feb 10 16&#8203;:27&#8203;:13 2016 \-0700
>
>     regcomp\.c&#8203;: Fix some parsing glitches
>
>     I undertook a code review of how regcomp\.c parses things in light of
the
>     tickets found by the fuzzer\,
>     https://rt-archive.perl.org/perl5/Ticket/Display.html?id=126546.&nbsp;  This commit is the
>     result of my efforts so far\.  I was not planning to push it now\, but
the
>     work found a couple of new error messages that should be raised\, and
>     this has to be done before the visible changes code freeze coming up
all
>     too soon\.  I will add test cases after that freeze\, including if to
see
>     that these changes fix all the observed issues\.
>
>     The audit was tedious\, and may have missed some things\.  Several
issues
>     occurred in multiple places\.  One is to not advance the parse by
>     UTF8SKIP appropriately; another is to subtract one byte from the parse
>     and assume that that is pointing to the beginning of the previous
>     character \(which under UTF\-8 it may not\)\.  Another is to assume that
>     that the pattern is a C string\, that there are no interior NULs in it\.
>     I also found unnecessary tests\, given that an SV always has a
>     terminating NUL\.

This is not correct\. Xs code can and in practice does create sv's with no
trailing nulls\.

> :040000 040000 a58a2d3154c5e346f9dfab14f7ca9897cbe06cfb
41b83ee5e555d3ddc87935c3a3b30db821741214 M      pod
> :100644 100644 c00fdffd93dd9de69326e9b567d912dd9a1bee3f
aa06bc6e8da12440f5d5b649cd5faa24a8853cc7 M      regcomp\.c
> bisect run success
>
>
>
> \*\*GDB\*\*
>
> dcollins@&#8203;nightshade64&#8203;:\~$ gdb \-\-args perldebug/miniperl \-e '/\(?\<=/'
> GNU gdb \(GDB\) 7\.10
> Copyright \(C\) 2015 Free Software Foundation\, Inc\.
> License GPLv3\+&#8203;: GNU GPL version 3 or later \<
http&#8203;://gnu\.org/licenses/gpl\.html>
> This is free software&#8203;: you are free to change and redistribute it\.
> There is NO WARRANTY\, to the extent permitted by law\.  Type "show copying"
> and "show warranty" for details\.
> This GDB was configured as "x86\_64\-unknown\-linux\-gnu"\.
> Type "show configuration" for configuration details\.
> For bug reporting instructions\, please see&#8203;:
> \<http&#8203;://www\.gnu\.org/software/gdb/bugs/>\.
> Find the GDB manual and other documentation resources online at&#8203;:
> \<http&#8203;://www\.gnu\.org/software/gdb/documentation/>\.
> For help\, type "help"\.
> Type "apropos word" to search for commands related to "word"\.\.\.
> Reading symbols from perldebug/miniperl\.\.\.done\.
> \(gdb\) run
> Starting program&#8203;: /home/dcollins/perldebug/miniperl \-e /\\\(\\?\\\<=/
> \[Thread debugging using libthread\_db enabled\]
> Using host libthread\_db library "/lib/x86\_64\-linux\-gnu/libthread\_db\.so\.1"\.
> miniperl&#8203;: regcomp\.c&#8203;:10621&#8203;: S\_reg&#8203;: Assertion \`\(pRExC\_state\->parse\) \<
\(pRExC\_state\->end\)' failed\.
>
> Program received signal SIGABRT\, Aborted\.
> 0x00007ffff6cf9478 in raise \(\) from /lib/x86\_64\-linux\-gnu/libc\.so\.6
> \(gdb\) bt
> \#0  0x00007ffff6cf9478 in raise \(\) from /lib/x86\_64\-linux\-gnu/libc\.so\.6
> \#1  0x00007ffff6cfa8fa in abort \(\) from /lib/x86\_64\-linux\-gnu/libc\.so\.6
> \#2  0x00007ffff6cf23a7 in ?? \(\) from /lib/x86\_64\-linux\-gnu/libc\.so\.6
> \#3  0x00007ffff6cf2452 in \_\_assert\_fail \(\) from
/lib/x86\_64\-linux\-gnu/libc\.so\.6
> \#4  0x00000000006efe09 in S\_reg \(pRExC\_state=0x7fffffffdcf0\, paren=60\,
flagp=0x7fffffffd8bc\, depth=5\) at regcomp\.c&#8203;:10621
> \#5  0x00000000007186ed in S\_regatom \(pRExC\_state=pRExC\_state@&#8203;entry=0x7fffffffdcf0\,
flagp=flagp@&#8203;entry=0x7fffffffda34\, depth=depth@&#8203;entry=4\) at regcomp\.c&#8203;:12354
> \#6  0x0000000000725d0d in S\_regpiece \(depth=\<optimized out>\,
flagp=\<synthetic pointer>\, pRExC\_state=0x7fffffffdcf0\) at regcomp\.c&#8203;:11421
> \#7  S\_regbranch \(pRExC\_state=pRExC\_state@&#8203;entry=0x7fffffffdcf0\,
flagp=flagp@&#8203;entry=0x7fffffffdacc\, first=first@&#8203;entry=1\, depth=depth@&#8203;entry=2\)
at regcomp\.c&#8203;:11346
> \#8  0x0000000000745097 in S\_reg \(pRExC\_state=pRExC\_state@&#8203;entry=0x7fffffffdcf0\,
flagp=flagp@&#8203;entry=0x7fffffffdbbc\, depth=1\, paren=0\) at regcomp\.c&#8203;:11084
> \#9  0x000000000077b231 in Perl\_re\_op\_compile \(patternp=\<optimized out>\,
pat\_count=\<optimized out>\, expr=\<optimized out>\, eng=0xf33160
\<PL\_core\_reg\_engine>\, old\_re=0x0\,
>     is\_bare\_re=\<optimized out>\, orig\_rx\_flags=0\, pm\_flags=0\) at
regcomp\.c&#8203;:7027
> \#10 0x00000000004c4033 in Perl\_pmruntime \(o=0x1202718\, expr=0x12026d8\,
repl=0x0\, isreg=isreg@&#8203;entry=true\, floor=\<optimized out>\) at op\.c&#8203;:5632
> \#11 0x000000000064b5cd in Perl\_yyparse \(gramtype=gramtype@&#8203;entry=258\) at
perly\.y&#8203;:1028
> \#12 0x0000000000512111 in S\_parse\_body \(xsinit=0xec1b00 \<xs\_init>\,
env=0x0\) at perl\.c&#8203;:2365
> \#13 perl\_parse \(my\_perl=\<optimized out>\, xsinit=xsinit@&#8203;entry=0xec1b00
\<xs\_init>\, argc=\<optimized out>\, argv=\<optimized out>\, env=env@&#8203;entry=0x0\)
at perl\.c&#8203;:1681
> \#14 0x000000000040f838 in main \(argc=\<optimized out>\, argv=\<optimized
out>\, env=\<optimized out>\) at miniperlmain\.c&#8203;:120
> \(gdb\) f 4
> \#4  0x00000000006efe09 in S\_reg \(pRExC\_state=0x7fffffffdcf0\, paren=60\,
flagp=0x7fffffffd8bc\, depth=5\) at regcomp\.c&#8203;:10621
> 10621                   assert\(RExC\_parse \< RExC\_end\);
> \(gdb\) info locals
> is\_logical = false
> seqstart = \<optimized out>
> endptr = 0x0
> has\_intervening\_patws = false
> ret = 0x0
> br = \<optimized out>
> lastbr = \<optimized out>
> ender = 0x0
> parno = 0
> flags = 0
> oregflags = 0
> have\_branch = false
> is\_open = false
> freeze\_paren = 0
> after\_freeze = 0
> num = \<optimized out>
> parse\_start = 0x1202861 "?\<="
> oregcomp\_parse = 0x1202861 "?\<="
> re\_debug\_flags = 0
> \(gdb\) l
> 10616                       goto capturing\_parens;
> 10617                   \}
> 10618                   RExC\_seen |= REG\_LOOKBEHIND\_SEEN;
> 10619                   RExC\_in\_lookbehind\+\+;
> 10620                   RExC\_parse\+\+;
> 10621                   assert\(RExC\_parse \< RExC\_end\);
> 10622                   /\* FALLTHROUGH \*/
> 10623               case '='&#8203;:           /\* \(?=\.\.\.\) \*/
> 10624                   RExC\_seen\_zerolen\+\+;
> 10625                   break;
> \(gdb\)
>
>
> \*\*PERL \-V\*\*
>
> dcollins@&#8203;nightshade64&#8203;:\~/perl$ \./perl \-Ilib \-V
> Summary of my perl5 \(revision 5 version 25 subversion 1\) configuration&#8203;:
>   Commit id&#8203;: 8255316ac82c0fbd7ae49f2f9a217ce063d7bbfb
>   Platform&#8203;:
>     osname=linux\, osvers=4\.5\.0\-2\-amd64\, archname=x86\_64\-linux\-ld
>     uname='linux nightshade64 4\.5\.0\-2\-amd64 \#1 smp debian 4\.5\.3\-2
\(2016\-05\-08\) x86\_64 gnulinux '
>     config\_args='\-Dusedevel \-Dprefix=/usr/local/perl\-afl \-Dcc=ccache
afl\-gcc \-Duselongdouble \-Duse64bitall \-Doptimize=\-g \-Uversiononly \-Uman1dir
\-Uman3dir \-des'
>     hint=recommended\, useposix=true\, d\_sigaction=define
>     useithreads=undef\, usemultiplicity=undef
>     use64bitint=define\, use64bitall=define\, uselongdouble=define
>     usemymalloc=n\, bincompat5005=undef
>   Compiler&#8203;:
>     cc='ccache afl\-gcc'\, ccflags ='\-fwrapv \-fno\-strict\-aliasing \-pipe
\-fstack\-protector\-strong \-I/usr/local/include \-D\_LARGEFILE\_SOURCE
\-D\_FILE\_OFFSET\_BITS=64'\,
>     optimize='\-g'\,
>     cppflags='\-fwrapv \-fno\-strict\-aliasing \-pipe \-fstack\-protector\-strong
\-I/usr/local/include'
>     ccversion=''\, gccversion='6\.1\.0'\, gccosandvers=''
>     intsize=4\, longsize=8\, ptrsize=8\, doublesize=8\, byteorder=12345678\,
doublekind=3
>     d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16\,
longdblkind=3
>     ivtype='long'\, ivsize=8\, nvtype='long double'\, nvsize=16\,
Off\_t='off\_t'\, lseeksize=8
>     alignbytes=16\, prototype=define
>   Linker and Libraries&#8203;:
>     ld='ccache afl\-gcc'\, ldflags =' \-fstack\-protector\-strong
\-L/usr/local/lib'
>     libpth=/usr/local/lib
/usr/local/lib/gcc/x86\_64\-pc\-linux\-gnu/6\.1\.0/include\-fixed
/usr/include/x86\_64\-linux\-gnu /usr/lib /lib/x86\_64\-linux\-gnu /lib/\.\./lib
/usr/lib/x86\_64\-linux\-gnu /usr/lib/\.\./lib /lib
>     libs=\-lpthread \-lnsl \-ldl \-lm \-lcrypt \-lutil \-lc
>     perllibs=\-lpthread \-lnsl \-ldl \-lm \-lcrypt \-lutil \-lc
>     libc=libc\-2\.22\.so\, so=so\, useshrplib=false\, libperl=libperl\.a
>     gnulibc\_version='2\.22'
>   Dynamic Linking&#8203;:
>     dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-Wl\,\-E'
>     cccdlflags='\-fPIC'\, lddlflags='\-shared \-g \-L/usr/local/lib
\-fstack\-protector\-strong'
>
>
> Characteristics of this binary \(from libperl\)&#8203;:
>   Compile\-time options&#8203;: HAS\_TIMES PERLIO\_LAYERS PERL\_COPY\_ON\_WRITE
>                         PERL\_DONT\_CREATE\_GVSV
>                         PERL\_HASH\_FUNC\_ONE\_AT\_A\_TIME\_HARD PERL\_MALLOC\_WRAP
>                         PERL\_OP\_PARENT PERL\_PRESERVE\_IVUV PERL\_USE\_DEVEL
>                         USE\_64\_BIT\_ALL USE\_64\_BIT\_INT USE\_LARGE\_FILES
>                         USE\_LOCALE USE\_LOCALE\_COLLATE USE\_LOCALE\_CTYPE
>                         USE\_LOCALE\_NUMERIC USE\_LOCALE\_TIME USE\_LONG\_DOUBLE
>                         USE\_PERLIO USE\_PERL\_ATOF
>   Built under linux
>   Compiled at May 16 2016 07&#8203;:45&#8203;:27
>   @&#8203;INC&#8203;:
>     lib
>     /usr/local/perl\-afl/lib/site\_perl/5\.25\.1/x86\_64\-linux\-ld
>     /usr/local/perl\-afl/lib/site\_perl/5\.25\.1
>     /usr/local/perl\-afl/lib/5\.25\.1/x86\_64\-linux\-ld
>     /usr/local/perl\-afl/lib/5\.25\.1
>     /usr/local/perl\-afl/lib/site\_perl/5\.24\.0
>     /usr/local/perl\-afl/lib/site\_perl
>     \.
>

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @khwilliamson

On 05/20/2016 10&#8203;:32 PM\, demerphq wrote&#8203;:
>
> On 17 May 2016 3&#8203;:04 p\.m\.\, "Dan Collins" \<perlbug\-followup@&#8203;perl\.org
> \<mailto&#8203;:perlbug\-followup@&#8203;perl\.org>> wrote&#8203;:
>  >
>  > \# New Ticket Created by  Dan Collins
>  > \# Please include the string&#8203;:  \[perl \#128170\]
>  > \# in the subject line of all future correspondence about this issue\.
>  > \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=128170 >
>  >
>  >
>  > Greetings Porters\,
>  >
>  > I have compiled bleadperl with the afl\-gcc compiler using&#8203;:
>  >
>  > \./Configure \-Dusedevel \-Dprefix='/usr/local/perl\-afl' \-Dcc='ccache
> afl\-gcc' \-Uuselongdouble \-Duse64bitall \-Doptimize=\-g \-Uversiononly
> \-Uman1dir \-Uman3dir \-Dusequadmath \-des
>  > AFL\_HARDEN=1 make && make test
>  >
>  > And then fuzzed the resulting binary using&#8203;:
>  >
>  > AFL\_NO\_VAR\_CHECK=1 afl\-fuzz \-i in \-o out bin/perl @&#8203;@&#8203;
>  >
>  > After reducing testcases using \`afl\-tmin\` and performing additional
> minimization by hand\, I have located the following testcase that
> triggers an assert fail in debugging builds of the perl interpreter\. The
> testcase is the 6\-character file&#8203;:
>  >
>  > /\(?\<=/
>  >
>  > On normal builds\, this exits with the expected error\. On debug
> builds\, this returns an assert fail\.
>  >
>  > dcollins@&#8203;nightshade64&#8203;:\~$ perl/miniperl \-e '/\(?\<=/'
>  > Unmatched \( in regex; marked by \<\-\- HERE in m/\( \<\-\- HERE ?\<=/ at \-e
> line 1\.
>  > dcollins@&#8203;nightshade64&#8203;:\~$ perldebug/miniperl \-e '/\(?\<=/'
>  > miniperl&#8203;: regcomp\.c&#8203;:10621&#8203;: S\_reg&#8203;: Assertion \`\(pRExC\_state\->parse\) \<
> \(pRExC\_state\->end\)' failed\.
>  > Aborted
>  >
>  > Debugging tool output is below\. A bisect was performed and points to
> the following diff&#8203;:
>  >
>  > cfbef7dc3bfb89e4ed2c00ea9c9e3bcfd0b170fd is the first bad commit
>  > commit cfbef7dc3bfb89e4ed2c00ea9c9e3bcfd0b170fd
>  > Author&#8203;: Karl Williamson \<khw@&#8203;cpan\.org \<mailto&#8203;:khw@&#8203;cpan\.org>>
>  > Date&#8203;:   Wed Feb 10 16&#8203;:27&#8203;:13 2016 \-0700
>  >
>  >     regcomp\.c&#8203;: Fix some parsing glitches
>  >
>  >     I undertook a code review of how regcomp\.c parses things in light
> of the
>  >     tickets found by the fuzzer\,
>  > https://rt-archive.perl.org/perl5/Ticket/Display.html?id=126546.&nbsp;  This commit is the
>  >     result of my efforts so far\.  I was not planning to push it now\,
> but the
>  >     work found a couple of new error messages that should be raised\, and
>  >     this has to be done before the visible changes code freeze coming
> up all
>  >     too soon\.  I will add test cases after that freeze\, including if
> to see
>  >     that these changes fix all the observed issues\.
>  >
>  >     The audit was tedious\, and may have missed some things\.  Several
> issues
>  >     occurred in multiple places\.  One is to not advance the parse by
>  >     UTF8SKIP appropriately; another is to subtract one byte from the
> parse
>  >     and assume that that is pointing to the beginning of the previous
>  >     character \(which under UTF\-8 it may not\)\.  Another is to assume that
>  >     that the pattern is a C string\, that there are no interior NULs
> in it\.
>  >     I also found unnecessary tests\, given that an SV always has a
>  >     terminating NUL\.
>
> This is not correct\. Xs code can and in practice does create sv's with
> no trailing nulls\.

Hmm\.  I've written a bunch of code with that assumption\, partly because 
a bunch of code that was already in place has that assumption\.

I did add assertions to this effect in the regex code at the same time 
as this commit was done\.

Can you give an example of how XS could do this?  All I can think of is 
if it arbitrarily turns on a POK flag\.
>
>  > :040000 040000 a58a2d3154c5e346f9dfab14f7ca9897cbe06cfb
> 41b83ee5e555d3ddc87935c3a3b30db821741214 M      pod
>  > :100644 100644 c00fdffd93dd9de69326e9b567d912dd9a1bee3f
> aa06bc6e8da12440f5d5b649cd5faa24a8853cc7 M      regcomp\.c
>  > bisect run success
>  >
>  >
>  >
>  > \*\*GDB\*\*
>  >
>  > dcollins@&#8203;nightshade64&#8203;:\~$ gdb \-\-args perldebug/miniperl \-e '/\(?\<=/'
>  > GNU gdb \(GDB\) 7\.10
>  > Copyright \(C\) 2015 Free Software Foundation\, Inc\.
>  > License GPLv3\+&#8203;: GNU GPL version 3 or later
> \<http&#8203;://gnu\.org/licenses/gpl\.html>
>  > This is free software&#8203;: you are free to change and redistribute it\.
>  > There is NO WARRANTY\, to the extent permitted by law\.  Type "show
> copying"
>  > and "show warranty" for details\.
>  > This GDB was configured as "x86\_64\-unknown\-linux\-gnu"\.
>  > Type "show configuration" for configuration details\.
>  > For bug reporting instructions\, please see&#8203;:
>  > \<http&#8203;://www\.gnu\.org/software/gdb/bugs/>\.
>  > Find the GDB manual and other documentation resources online at&#8203;:
>  > \<http&#8203;://www\.gnu\.org/software/gdb/documentation/>\.
>  > For help\, type "help"\.
>  > Type "apropos word" to search for commands related to "word"\.\.\.
>  > Reading symbols from perldebug/miniperl\.\.\.done\.
>  > \(gdb\) run
>  > Starting program&#8203;: /home/dcollins/perldebug/miniperl \-e /\\\(\\?\\\<=/
>  > \[Thread debugging using libthread\_db enabled\]
>  > Using host libthread\_db library
> "/lib/x86\_64\-linux\-gnu/libthread\_db\.so\.1"\.
>  > miniperl&#8203;: regcomp\.c&#8203;:10621&#8203;: S\_reg&#8203;: Assertion \`\(pRExC\_state\->parse\) \<
> \(pRExC\_state\->end\)' failed\.
>  >
>  > Program received signal SIGABRT\, Aborted\.
>  > 0x00007ffff6cf9478 in raise \(\) from /lib/x86\_64\-linux\-gnu/libc\.so\.6
>  > \(gdb\) bt
>  > \#0  0x00007ffff6cf9478 in raise \(\) from /lib/x86\_64\-linux\-gnu/libc\.so\.6
>  > \#1  0x00007ffff6cfa8fa in abort \(\) from /lib/x86\_64\-linux\-gnu/libc\.so\.6
>  > \#2  0x00007ffff6cf23a7 in ?? \(\) from /lib/x86\_64\-linux\-gnu/libc\.so\.6
>  > \#3  0x00007ffff6cf2452 in \_\_assert\_fail \(\) from
> /lib/x86\_64\-linux\-gnu/libc\.so\.6
>  > \#4  0x00000000006efe09 in S\_reg \(pRExC\_state=0x7fffffffdcf0\,
> paren=60\, flagp=0x7fffffffd8bc\, depth=5\) at regcomp\.c&#8203;:10621
>  > \#5  0x00000000007186ed in S\_regatom
> \(pRExC\_state=pRExC\_state@&#8203;entry=0x7fffffffdcf0\,
> flagp=flagp@&#8203;entry=0x7fffffffda34\, depth=depth@&#8203;entry=4\) at regcomp\.c&#8203;:12354
>  > \#6  0x0000000000725d0d in S\_regpiece \(depth=\<optimized out>\,
> flagp=\<synthetic pointer>\, pRExC\_state=0x7fffffffdcf0\) at regcomp\.c&#8203;:11421
>  > \#7  S\_regbranch \(pRExC\_state=pRExC\_state@&#8203;entry=0x7fffffffdcf0\,
> flagp=flagp@&#8203;entry=0x7fffffffdacc\, first=first@&#8203;entry=1\,
> depth=depth@&#8203;entry=2\) at regcomp\.c&#8203;:11346
>  > \#8  0x0000000000745097 in S\_reg
> \(pRExC\_state=pRExC\_state@&#8203;entry=0x7fffffffdcf0\,
> flagp=flagp@&#8203;entry=0x7fffffffdbbc\, depth=1\, paren=0\) at regcomp\.c&#8203;:11084
>  > \#9  0x000000000077b231 in Perl\_re\_op\_compile \(patternp=\<optimized
> out>\, pat\_count=\<optimized out>\, expr=\<optimized out>\, eng=0xf33160
> \<PL\_core\_reg\_engine>\, old\_re=0x0\,
>  >     is\_bare\_re=\<optimized out>\, orig\_rx\_flags=0\, pm\_flags=0\) at
> regcomp\.c&#8203;:7027
>  > \#10 0x00000000004c4033 in Perl\_pmruntime \(o=0x1202718\,
> expr=0x12026d8\, repl=0x0\, isreg=isreg@&#8203;entry=true\, floor=\<optimized out>\)
> at op\.c&#8203;:5632
>  > \#11 0x000000000064b5cd in Perl\_yyparse \(gramtype=gramtype@&#8203;entry=258\)
> at perly\.y&#8203;:1028
>  > \#12 0x0000000000512111 in S\_parse\_body \(xsinit=0xec1b00 \<xs\_init>\,
> env=0x0\) at perl\.c&#8203;:2365
>  > \#13 perl\_parse \(my\_perl=\<optimized out>\, xsinit=xsinit@&#8203;entry=0xec1b00
> \<xs\_init>\, argc=\<optimized out>\, argv=\<optimized out>\,
> env=env@&#8203;entry=0x0\) at perl\.c&#8203;:1681
>  > \#14 0x000000000040f838 in main \(argc=\<optimized out>\, argv=\<optimized
> out>\, env=\<optimized out>\) at miniperlmain\.c&#8203;:120
>  > \(gdb\) f 4
>  > \#4  0x00000000006efe09 in S\_reg \(pRExC\_state=0x7fffffffdcf0\,
> paren=60\, flagp=0x7fffffffd8bc\, depth=5\) at regcomp\.c&#8203;:10621
>  > 10621                   assert\(RExC\_parse \< RExC\_end\);
>  > \(gdb\) info locals
>  > is\_logical = false
>  > seqstart = \<optimized out>
>  > endptr = 0x0
>  > has\_intervening\_patws = false
>  > ret = 0x0
>  > br = \<optimized out>
>  > lastbr = \<optimized out>
>  > ender = 0x0
>  > parno = 0
>  > flags = 0
>  > oregflags = 0
>  > have\_branch = false
>  > is\_open = false
>  > freeze\_paren = 0
>  > after\_freeze = 0
>  > num = \<optimized out>
>  > parse\_start = 0x1202861 "?\<="
>  > oregcomp\_parse = 0x1202861 "?\<="
>  > re\_debug\_flags = 0
>  > \(gdb\) l
>  > 10616                       goto capturing\_parens;
>  > 10617                   \}
>  > 10618                   RExC\_seen |= REG\_LOOKBEHIND\_SEEN;
>  > 10619                   RExC\_in\_lookbehind\+\+;
>  > 10620                   RExC\_parse\+\+;
>  > 10621                   assert\(RExC\_parse \< RExC\_end\);
>  > 10622                   /\* FALLTHROUGH \*/
>  > 10623               case '='&#8203;:           /\* \(?=\.\.\.\) \*/
>  > 10624                   RExC\_seen\_zerolen\+\+;
>  > 10625                   break;
>  > \(gdb\)
>  >
>  >
>  > \*\*PERL \-V\*\*
>  >
>  > dcollins@&#8203;nightshade64&#8203;:\~/perl$ \./perl \-Ilib \-V
>  > Summary of my perl5 \(revision 5 version 25 subversion 1\) configuration&#8203;:
>  >   Commit id&#8203;: 8255316ac82c0fbd7ae49f2f9a217ce063d7bbfb
>  >   Platform&#8203;:
>  >     osname=linux\, osvers=4\.5\.0\-2\-amd64\, archname=x86\_64\-linux\-ld
>  >     uname='linux nightshade64 4\.5\.0\-2\-amd64 \#1 smp debian 4\.5\.3\-2
> \(2016\-05\-08\) x86\_64 gnulinux '
>  >     config\_args='\-Dusedevel \-Dprefix=/usr/local/perl\-afl \-Dcc=ccache
> afl\-gcc \-Duselongdouble \-Duse64bitall \-Doptimize=\-g \-Uversiononly
> \-Uman1dir \-Uman3dir \-des'
>  >     hint=recommended\, useposix=true\, d\_sigaction=define
>  >     useithreads=undef\, usemultiplicity=undef
>  >     use64bitint=define\, use64bitall=define\, uselongdouble=define
>  >     usemymalloc=n\, bincompat5005=undef
>  >   Compiler&#8203;:
>  >     cc='ccache afl\-gcc'\, ccflags ='\-fwrapv \-fno\-strict\-aliasing \-pipe
> \-fstack\-protector\-strong \-I/usr/local/include \-D\_LARGEFILE\_SOURCE
> \-D\_FILE\_OFFSET\_BITS=64'\,
>  >     optimize='\-g'\,
>  >     cppflags='\-fwrapv \-fno\-strict\-aliasing \-pipe
> \-fstack\-protector\-strong \-I/usr/local/include'
>  >     ccversion=''\, gccversion='6\.1\.0'\, gccosandvers=''
>  >     intsize=4\, longsize=8\, ptrsize=8\, doublesize=8\,
> byteorder=12345678\, doublekind=3
>  >     d\_longlong=define\, longlongsize=8\, d\_longdbl=define\,
> longdblsize=16\, longdblkind=3
>  >     ivtype='long'\, ivsize=8\, nvtype='long double'\, nvsize=16\,
> Off\_t='off\_t'\, lseeksize=8
>  >     alignbytes=16\, prototype=define
>  >   Linker and Libraries&#8203;:
>  >     ld='ccache afl\-gcc'\, ldflags =' \-fstack\-protector\-strong
> \-L/usr/local/lib'
>  >     libpth=/usr/local/lib
> /usr/local/lib/gcc/x86\_64\-pc\-linux\-gnu/6\.1\.0/include\-fixed
> /usr/include/x86\_64\-linux\-gnu /usr/lib /lib/x86\_64\-linux\-gnu /lib/\.\./lib
> /usr/lib/x86\_64\-linux\-gnu /usr/lib/\.\./lib /lib
>  >     libs=\-lpthread \-lnsl \-ldl \-lm \-lcrypt \-lutil \-lc
>  >     perllibs=\-lpthread \-lnsl \-ldl \-lm \-lcrypt \-lutil \-lc
>  >     libc=libc\-2\.22\.so \<http&#8203;://libc\-2\.22\.so>\, so=so\, useshrplib=false\,
> libperl=libperl\.a
>  >     gnulibc\_version='2\.22'
>  >   Dynamic Linking&#8203;:
>  >     dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-Wl\,\-E'
>  >     cccdlflags='\-fPIC'\, lddlflags='\-shared \-g \-L/usr/local/lib
> \-fstack\-protector\-strong'
>  >
>  >
>  > Characteristics of this binary \(from libperl\)&#8203;:
>  >   Compile\-time options&#8203;: HAS\_TIMES PERLIO\_LAYERS PERL\_COPY\_ON\_WRITE
>  >                         PERL\_DONT\_CREATE\_GVSV
>  >                         PERL\_HASH\_FUNC\_ONE\_AT\_A\_TIME\_HARD
> PERL\_MALLOC\_WRAP
>  >                         PERL\_OP\_PARENT PERL\_PRESERVE\_IVUV PERL\_USE\_DEVEL
>  >                         USE\_64\_BIT\_ALL USE\_64\_BIT\_INT USE\_LARGE\_FILES
>  >                         USE\_LOCALE USE\_LOCALE\_COLLATE USE\_LOCALE\_CTYPE
>  >                         USE\_LOCALE\_NUMERIC USE\_LOCALE\_TIME
> USE\_LONG\_DOUBLE
>  >                         USE\_PERLIO USE\_PERL\_ATOF
>  >   Built under linux
>  >   Compiled at May 16 2016 07&#8203;:45&#8203;:27
>  >   @&#8203;INC&#8203;:
>  >     lib
>  >     /usr/local/perl\-afl/lib/site\_perl/5\.25\.1/x86\_64\-linux\-ld
>  >     /usr/local/perl\-afl/lib/site\_perl/5\.25\.1
>  >     /usr/local/perl\-afl/lib/5\.25\.1/x86\_64\-linux\-ld
>  >     /usr/local/perl\-afl/lib/5\.25\.1
>  >     /usr/local/perl\-afl/lib/site\_perl/5\.24\.0
>  >     /usr/local/perl\-afl/lib/site\_perl
>  >     \.
>  >
>

**Comment by p5pRT** [comment]
### From @demerphq

On 23 May 2016 16&#8203;:06\, "Karl Williamson" \<public@&#8203;khwilliamson\.com> wrote&#8203;:
>
> On 05/20/2016 10&#8203;:32 PM\, demerphq wrote&#8203;:
>>
>>
>> On 17 May 2016 3&#8203;:04 p\.m\.\, "Dan Collins" \<perlbug\-followup@&#8203;perl\.org
>> \<mailto&#8203;:perlbug\-followup@&#8203;perl\.org>> wrote&#8203;:
>>  >
>>  > \# New Ticket Created by  Dan Collins
>>  > \# Please include the string&#8203;:  \[perl \#128170\]
>>  > \# in the subject line of all future correspondence about this issue\.
>>  > \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=128170 >
>>  >
>>  >
>>  > Greetings Porters\,
>>  >
>>  > I have compiled bleadperl with the afl\-gcc compiler using&#8203;:
>>  >
>>  > \./Configure \-Dusedevel \-Dprefix='/usr/local/perl\-afl' \-Dcc='ccache
>> afl\-gcc' \-Uuselongdouble \-Duse64bitall \-Doptimize=\-g \-Uversiononly
>> \-Uman1dir \-Uman3dir \-Dusequadmath \-des
>>  > AFL\_HARDEN=1 make && make test
>>  >
>>  > And then fuzzed the resulting binary using&#8203;:
>>  >
>>  > AFL\_NO\_VAR\_CHECK=1 afl\-fuzz \-i in \-o out bin/perl @&#8203;@&#8203;
>>  >
>>  > After reducing testcases using \`afl\-tmin\` and performing additional
>> minimization by hand\, I have located the following testcase that
>> triggers an assert fail in debugging builds of the perl interpreter\. The
>> testcase is the 6\-character file&#8203;:
>>  >
>>  > /\(?\<=/
>>  >
>>  > On normal builds\, this exits with the expected error\. On debug
>> builds\, this returns an assert fail\.
>>  >
>>  > dcollins@&#8203;nightshade64&#8203;:\~$ perl/miniperl \-e '/\(?\<=/'
>>  > Unmatched \( in regex; marked by \<\-\- HERE in m/\( \<\-\- HERE ?\<=/ at \-e
>> line 1\.
>>  > dcollins@&#8203;nightshade64&#8203;:\~$ perldebug/miniperl \-e '/\(?\<=/'
>>  > miniperl&#8203;: regcomp\.c&#8203;:10621&#8203;: S\_reg&#8203;: Assertion \`\(pRExC\_state\->parse\) \<
>> \(pRExC\_state\->end\)' failed\.
>>  > Aborted
>>  >
>>  > Debugging tool output is below\. A bisect was performed and points to
>> the following diff&#8203;:
>>  >
>>  > cfbef7dc3bfb89e4ed2c00ea9c9e3bcfd0b170fd is the first bad commit
>>  > commit cfbef7dc3bfb89e4ed2c00ea9c9e3bcfd0b170fd
>>  > Author&#8203;: Karl Williamson \<khw@&#8203;cpan\.org \<mailto&#8203;:khw@&#8203;cpan\.org>>
>>
>>  > Date&#8203;:   Wed Feb 10 16&#8203;:27&#8203;:13 2016 \-0700
>>  >
>>  >     regcomp\.c&#8203;: Fix some parsing glitches
>>  >
>>  >     I undertook a code review of how regcomp\.c parses things in light
>> of the
>>  >     tickets found by the fuzzer\,
>>  > https://rt-archive.perl.org/perl5/Ticket/Display.html?id=126546.&nbsp;  This commit is the
>>  >     result of my efforts so far\.  I was not planning to push it now\,
>> but the
>>  >     work found a couple of new error messages that should be raised\,
and
>>  >     this has to be done before the visible changes code freeze coming
>> up all
>>  >     too soon\.  I will add test cases after that freeze\, including if
>> to see
>>  >     that these changes fix all the observed issues\.
>>  >
>>  >     The audit was tedious\, and may have missed some things\.  Several
>> issues
>>  >     occurred in multiple places\.  One is to not advance the parse by
>>  >     UTF8SKIP appropriately; another is to subtract one byte from the
>> parse
>>  >     and assume that that is pointing to the beginning of the previous
>>  >     character \(which under UTF\-8 it may not\)\.  Another is to assume
that
>>  >     that the pattern is a C string\, that there are no interior NULs
>> in it\.
>>  >     I also found unnecessary tests\, given that an SV always has a
>>  >     terminating NUL\.
>>
>> This is not correct\. Xs code can and in practice does create sv's with
>> no trailing nulls\.
>
>
> Hmm\.  I've written a bunch of code with that assumption\, partly because a
bunch of code that was already in place has that assumption\.
>
> I did add assertions to this effect in the regex code at the same time as
this commit was done\.
>
> Can you give an example of how XS could do this?  All I can think of is
if it arbitrarily turns on a POK flag\.

Yeah\, exactly\. But in xs that is pretty standard\. I wouldnt rely on our
strings always having a trailing null\.

Yves

**Comment by p5pRT** [comment]
### From @cpansprout

On Mon May 23 14&#8203;:16&#8203;:31 2016\, demerphq wrote&#8203;:
> Yeah\, exactly\. But in xs that is pretty standard\. I wouldnt rely on our
> strings always having a trailing null\.

In fact\, I would like to remove that requirement \(to allow substr to avoid copying the string\)\, but someone else has pointed out that it is also common in XS to pass SvPV to system functions that expect a trailing null\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @demerphq

On 23 May 2016 6&#8203;:46 p\.m\.\, "Father Chrysostomos via RT" \<
perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
>
> On Mon May 23 14&#8203;:16&#8203;:31 2016\, demerphq wrote&#8203;:
> > Yeah\, exactly\. But in xs that is pretty standard\. I wouldnt rely on our
> > strings always having a trailing null\.
>
> In fact\, I would like to remove that requirement \(to allow substr to
avoid copying the string\)\, but someone else has pointed out that it is also
common in XS to pass SvPV to system functions that expect a trailing null\.

Maybe we should have a SvPV\_null macro that DTRT for such cases and abandon
any requirements that pv's are null terminated\.

Yves

**Comment by p5pRT** [comment]
### From @cpansprout

On Mon May 23 16&#8203;:07&#8203;:03 2016\, demerphq wrote&#8203;:
> On 23 May 2016 6&#8203;:46 p\.m\.\, "Father Chrysostomos via RT" \<
> perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> >
> > On Mon May 23 14&#8203;:16&#8203;:31 2016\, demerphq wrote&#8203;:
> > > Yeah\, exactly\. But in xs that is pretty standard\. I wouldnt rely on our
> > > strings always having a trailing null\.
> >
> > In fact\, I would like to remove that requirement \(to allow substr to
> avoid copying the string\)\, but someone else has pointed out that it is also
> common in XS to pass SvPV to system functions that expect a trailing null\.
> 
> Maybe we should have a SvPV\_null macro that DTRT for such cases and abandon
> any requirements that pv's are null terminated\.

It might have to be the other way round \(SvPV guarantees a null; a new variant returns the actual string\)\, for backward\-compatibility\.  I have opened a separate ticket for this \(\#128226\)\, since it is not directly related to the topic of this ticket\.


\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @khwilliamson

On 05/23/2016 03&#8203;:15 PM\, demerphq wrote&#8203;:
>
> On 23 May 2016 16&#8203;:06\, "Karl Williamson" \<public@&#8203;khwilliamson\.com
> \<mailto&#8203;:public@&#8203;khwilliamson\.com>> wrote&#8203;:
>  >
>  > On 05/20/2016 10&#8203;:32 PM\, demerphq wrote&#8203;:
>  >>
>  >>
>  >> On 17 May 2016 3&#8203;:04 p\.m\.\, "Dan Collins" \<perlbug\-followup@&#8203;perl\.org
> \<mailto&#8203;:perlbug\-followup@&#8203;perl\.org>
>  >> \<mailto&#8203;:perlbug\-followup@&#8203;perl\.org
> \<mailto&#8203;:perlbug\-followup@&#8203;perl\.org>>> wrote&#8203;:
>  >>  >
>  >>  > \# New Ticket Created by  Dan Collins
>  >>  > \# Please include the string&#8203;:  \[perl \#128170\]
>  >>  > \# in the subject line of all future correspondence about this issue\.
>  >>  > \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=128170 >
>  >>  >
>  >>  >
>  >>  > Greetings Porters\,
>  >>  >
>  >>  > I have compiled bleadperl with the afl\-gcc compiler using&#8203;:
>  >>  >
>  >>  > \./Configure \-Dusedevel \-Dprefix='/usr/local/perl\-afl' \-Dcc='ccache
>  >> afl\-gcc' \-Uuselongdouble \-Duse64bitall \-Doptimize=\-g \-Uversiononly
>  >> \-Uman1dir \-Uman3dir \-Dusequadmath \-des
>  >>  > AFL\_HARDEN=1 make && make test
>  >>  >
>  >>  > And then fuzzed the resulting binary using&#8203;:
>  >>  >
>  >>  > AFL\_NO\_VAR\_CHECK=1 afl\-fuzz \-i in \-o out bin/perl @&#8203;@&#8203;
>  >>  >
>  >>  > After reducing testcases using \`afl\-tmin\` and performing additional
>  >> minimization by hand\, I have located the following testcase that
>  >> triggers an assert fail in debugging builds of the perl interpreter\. The
>  >> testcase is the 6\-character file&#8203;:
>  >>  >
>  >>  > /\(?\<=/
>  >>  >
>  >>  > On normal builds\, this exits with the expected error\. On debug
>  >> builds\, this returns an assert fail\.
>  >>  >
>  >>  > dcollins@&#8203;nightshade64&#8203;:\~$ perl/miniperl \-e '/\(?\<=/'
>  >>  > Unmatched \( in regex; marked by \<\-\- HERE in m/\( \<\-\- HERE ?\<=/ at \-e
>  >> line 1\.
>  >>  > dcollins@&#8203;nightshade64&#8203;:\~$ perldebug/miniperl \-e '/\(?\<=/'
>  >>  > miniperl&#8203;: regcomp\.c&#8203;:10621&#8203;: S\_reg&#8203;: Assertion \`\(pRExC\_state\->parse\) \<
>  >> \(pRExC\_state\->end\)' failed\.
>  >>  > Aborted
>  >>  >
>  >>  > Debugging tool output is below\. A bisect was performed and points to
>  >> the following diff&#8203;:
>  >>  >
>  >>  > cfbef7dc3bfb89e4ed2c00ea9c9e3bcfd0b170fd is the first bad commit
>  >>  > commit cfbef7dc3bfb89e4ed2c00ea9c9e3bcfd0b170fd
>  >>  > Author&#8203;: Karl Williamson \<khw@&#8203;cpan\.org \<mailto&#8203;:khw@&#8203;cpan\.org>
> \<mailto&#8203;:khw@&#8203;cpan\.org \<mailto&#8203;:khw@&#8203;cpan\.org>>>
>  >>
>  >>  > Date&#8203;:   Wed Feb 10 16&#8203;:27&#8203;:13 2016 \-0700
>  >>  >
>  >>  >     regcomp\.c&#8203;: Fix some parsing glitches
>  >>  >
>  >>  >     I undertook a code review of how regcomp\.c parses things in light
>  >> of the
>  >>  >     tickets found by the fuzzer\,
>  >>  > https://rt-archive.perl.org/perl5/Ticket/Display.html?id=126546.&nbsp;  This commit
> is the
>  >>  >     result of my efforts so far\.  I was not planning to push it now\,
>  >> but the
>  >>  >     work found a couple of new error messages that should be
> raised\, and
>  >>  >     this has to be done before the visible changes code freeze coming
>  >> up all
>  >>  >     too soon\.  I will add test cases after that freeze\, including if
>  >> to see
>  >>  >     that these changes fix all the observed issues\.
>  >>  >
>  >>  >     The audit was tedious\, and may have missed some things\.  Several
>  >> issues
>  >>  >     occurred in multiple places\.  One is to not advance the parse by
>  >>  >     UTF8SKIP appropriately; another is to subtract one byte from the
>  >> parse
>  >>  >     and assume that that is pointing to the beginning of the previous
>  >>  >     character \(which under UTF\-8 it may not\)\.  Another is to
> assume that
>  >>  >     that the pattern is a C string\, that there are no interior NULs
>  >> in it\.
>  >>  >     I also found unnecessary tests\, given that an SV always has a
>  >>  >     terminating NUL\.
>  >>
>  >> This is not correct\. Xs code can and in practice does create sv's with
>  >> no trailing nulls\.
>  >
>  >
>  > Hmm\.  I've written a bunch of code with that assumption\, partly
> because a bunch of code that was already in place has that assumption\.
>  >
>  > I did add assertions to this effect in the regex code at the same
> time as this commit was done\.
>  >
>  > Can you give an example of how XS could do this?  All I can think of
> is if it arbitrarily turns on a POK flag\.
>
> Yeah\, exactly\. But in xs that is pretty standard\. I wouldnt rely on our
> strings always having a trailing null\.
>
> Yves
>

I'm having trouble understanding how one would set such a thing up\.  At 
some point one would have to populate a sequence of bytes\.  I don't see 
how to populate anything but a short sequence unless the SV is already a 
PV\.  Could you outline roughly how it would be done?

**Comment by p5pRT** [comment]
### From @cpansprout

On Mon May 23 21&#8203;:18&#8203;:16 2016\, public@&#8203;khwilliamson\.com wrote&#8203;:
> I'm having trouble understanding how one would set such a thing up\.
> At
> some point one would have to populate a sequence of bytes\.  I don't
> see
> how to populate anything but a short sequence unless the SV is already
> a
> PV\.  Could you outline roughly how it would be done?

http&#8203;://grep\.cpan\.me/?q=SvPVX\.\*\+%3D\[\+\]

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @geeknik

While fuzzing Perl v5\.25\.3 \(v5\.25\.2\-65\-g3866075\) with American Fuzzy Lop\, it was found that perl \-e '/\(?\<\!/' triggers the following assertion failure&#8203;:

perl&#8203;: regcomp\.c&#8203;:10636&#8203;: S\_reg&#8203;: Assertion \`\(pRExC\_state\->parse\) \< \(pRExC\_state\->end\)' failed\.
Aborted

**Comment by p5pRT** [comment]
### From @dcollinsn

Duplicate of RT \#128170\, with a bisect and some discussion there\. \(Slightly different code \- it seems both positive and negative assertions trigger this\.\) Merging\.

**Comment by p5pRT** [comment]
### From @khwilliamson

On 07/05/2016 09&#8203;:27 PM\, Brian Carpenter \(via RT\) wrote&#8203;:
> \# New Ticket Created by  Brian Carpenter
> \# Please include the string&#8203;:  \[perl \#128555\]
> \# in the subject line of all future correspondence about this issue\.
> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=128555 >
>
>
> While fuzzing Perl v5\.25\.3 \(v5\.25\.2\-65\-g3866075\) with American Fuzzy Lop\, it was found that perl \-e '/\(?\<\!/' triggers the following assertion failure&#8203;:
>
> perl&#8203;: regcomp\.c&#8203;:10636&#8203;: S\_reg&#8203;: Assertion \`\(pRExC\_state\->parse\) \< \(pRExC\_state\->end\)' failed\.
> Aborted
>
>

I will look into it

**Comment by p5pRT** [comment]
### From @khwilliamson

Fixed by f2e32b2c677945ce6a36531a7cd5696dc1ca2cef
\-\- 
Karl Williamson

**Comment by p5pRT** [comment]
@khwilliamson - Status changed from 'open' to 'pending release'

**Comment by p5pRT** [comment]
### From @khwilliamson

Thank you for filing this report\.  You have helped make Perl better\.

With the release today of Perl 5\.26\.0\, this and 210 other issues have been
resolved\.

Perl 5\.26\.0 may be downloaded via&#8203;:
https://metacpan.org/release/XSAWYERX/perl-5.26.0

If you find that the problem persists\, feel free to reopen this ticket\.

**Comment by p5pRT** [comment]
@khwilliamson - Status changed from 'pending release' to 'resolved'
